### PR TITLE
Create test_cpdAssignUStarTh20100901.m

### LINKED ISSUE #27: Generate unit tests for cpdAssignUStarTh20100901.m
Create a suite of unit tests for matlab function cpdAssignUStarTh20100901.m

Tian to create infrastructure

James to generate test data

### LINKED ISSUE #48: None vs NaN
We had a discussion earlier about how to deal with `NaN` values. In the JSON encoding, `NaN` is getting turned into `None`. We then have code that converts these back to `NaN` (https://github.com/Cambridge-ICCS/ONEFlux/blob/27-generate-unit-tests-for-cpdassignustar/tests/conftest.py#L73) however we still have some places, e.g., here: https://github.com/Cambridge-ICCS/ONEFlux/blob/27-generate-unit-tests-for-cpdassignustar/tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py#L302 where `None` is used. Are there any others?

My opinion is that we should preserve `NaN` where they appear, but we also need to remember that standard equality says that `NaN == NaN` is false so we need specialise equality routines if we expect NaNs...

### PR DESCRIPTION
**Pull request by j-emberton**
Closes #27 

Added unit tests for `cpdAssignUStarTh20100901` and updated `conftest.py` with a `MFWrapper` to wrap the matlab function. With `MFWrapper`, any function call by `matlab_engine.FUNC` will automatically handle the JSON encoding and decoding of non-scalar structs, as long as in the Python side we add keyword arguments `jsonencode=[...]` or `jsondecode=[...]` to `matlab_engine.FUNC()`. The wrapper also will show the printing in MATLAB functions at the end of the testing.

### CODE REVIEW

**Comment by dorchard** [comment]
Quite a few of the tests fail for me. Some seem to be to do with the creating of a struct to pass to MATLAB:
```
FAILED tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py::test_cpdAssignUStarTh20100901_basic - AttributeError: 'dict' object has no attribute 'mt'
FAILED tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py::test_cpdAssignUStarTh20100901_plotting - AttributeError: 'dict' object has no attribute 'mt'
```
Others seem to be to do with the matlab engine setup. 

I can't see why this would be a local problem for me in the latter case.

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    # Call MATLAB function
+    results = matlab_engine.cpdAssignUStarTh20100901(invalid_stats, fPlot, cSiteYr, nargout=11)
+
+    # Check if function handles invalid input gracefully
+    assert len(results[0]) == 0, "CpA should be empty for invalid input"
+    assert len(results[5]) > 0, "cFailure should contain an error message for invalid input"
+
+def test_cpdAssignUStarTh20100901_edge_cases(matlab_engine, mock_stats):
+    # Test edge cases (e.g., all significant change points, no significant change points)
+    edge_stats = matlab.engine.struct()
```
Comment (dorchard) [inline tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: This doesn't seem right to me. Shouldn't it be mathlab_engine.struct() since we need the runtime object to create us the struct?

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            self.cic2 = np.random.rand(40)
+            self.p = np.random.rand(40)
+
+    return MockStats()
+
+def test_cpdAssignUStarTh20100901_basic(matlab_engine, mock_stats):
+    # Convert mock stats to MATLAB structure
+    stats_matlab = matlab_engine.struct()
+    for field in ['mt', 'Cp', 'b1', 'c2', 'cib1', 'cic2', 'p']:
+        setattr(stats_matlab, field, matlab.double(getattr(mock_stats, field).tolist()))
```
Comment (dorchard) [inline tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: This pattern doesn't seem right to me: a MATLAB struct seems to be represented as a dictionary so using `setattr` doesn't work for me with R2024a (24.1):
```
>>> import matlab.engine
>>> s = eng.struct()
>>> setattr(s, "field", 42)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'dict' object has no attribute 'field'
```
Is this a change with R2024b? Do we need to change our requirements?

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    assert isinstance(cFailure, str), "cFailure should be a string"
+    assert isinstance(fSelect, matlab.logical), "fSelect should be a MATLAB logical array"
+    assert isinstance(sSine, matlab.double), "sSine should be a MATLAB double array"
+    assert isinstance(FracSig, float), "FracSig should be a float"
+    assert isinstance(FracModeD, float), "FracModeD should be a float"
+    assert isinstance(FracSelect, float), "FracSelect should be a float"
+
+def test_cpdAssignUStarTh20100901_plotting(matlab_engine, mock_stats):
+    # Test with plotting enabled
+    stats_matlab = matlab_engine.struct()
```
Comment (dorchard) [inline tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Doesn't work for me (Matlab runtime engine R2024a).

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    # Call MATLAB function
+    results = matlab_engine.cpdAssignUStarTh20100901(stats_matlab, fPlot, cSiteYr, nargout=11)
+
+    # Check if the function runs without errors when plotting is enabled
+    assert len(results) == 11, "Function should return 11 outputs even with plotting enabled"
+
+@pytest.mark.parametrize("invalid_input", [[], None, "invalid"])
+def test_cpdAssignUStarTh20100901_invalid_input(matlab_engine, invalid_input):
+    # Test with invalid input
+    invalid_stats = matlab.engine.struct()
```
Comment (dorchard) [inline tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Similar issue to other lines like this.

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py (hunk 2 above)
Comment (casterkay) [reply tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Fixed

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py (hunk 4 above)
Comment (casterkay) [reply tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Fixed

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py (hunk 3 above)
Comment (casterkay) [reply tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Fixed

File: tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py (hunk 1 above)
Comment (casterkay) [reply tests/unit_tests/test_ustar_cp/test_cpdAssignUStarTh20100901.py]: Fixed

**Comment by dorchard** [comment]
Thanks @tztsai  - I'm having a problem with the tests here `tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py::test_cpdBootstrap_against_testcases` seems to be hanging but I can't see why. I think it must be a local problem (two tests run then things get stuck).

**Comment by dorchard** [comment]
I think I've narrowed it down to tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py on line 143... which is strange.

**Comment by casterkay** [comment]
Now the handling of NaN and None values and their comparison are fixed. All tests pass.

**Comment by dorchard** [comment]
@tztsai Thanks! looks good to me. Could you resolve the conflicts with the target branch then I am happy to approve for merge.

**Comment by casterkay** [comment]
> @tztsai Thanks! looks good to me. Could you resolve the conflicts with the target branch then I am happy to approve for merge.

Resolved now

**Review by dorchard — APPROVED** [review]
